include/tvm/expr_operator.h

```diff
@@ -551,6 +557,12 @@ inline Expr MakeConstScalar(Type t, ValueType value) {
   if (t.is_int()) return ir::IntImm::make(t, static_cast<int64_t>(value));
   if (t.is_uint()) return ir::UIntImm::make(t, static_cast<uint64_t>(value));
   if (t.is_float()) return ir::FloatImm::make(t, static_cast<double>(value));
+  // For now, we store const scalar values of custom datatypes within doubles; later, during the
+  // datatypes lowering pass, we will lower the value to its true representation in the format
+  // specified by the datatype.
+  // TODO(gus) when do we need to start worrying about doubles not being precise enough?
```

Comment:
what are the chances that types we'll be exploring will be more precise than doubles? in that case we'd need to employ some sort of arbitrary precision data type library. I'd remove the TODO and make a note in the documentation/README so people are aware that this is our default type for storing values.

Reply:
Yep, we had that same problem in the posit library. Posits can be a LOT more precise than even 128-bit quad precision floats and if you use simple marshal casting then you will limit your number system universe badly. All compiler environments (C++, Go, etc.) use an arbitrary precision engine to manipulate constexpr. I would expect a compiler like TVM to have similar behavior, certainly not expect it to inject all the weirdness of IEEE double precision rounding.

Reply:
Good point; the question now is: should we address this limitation in this PR, or in a future PR? This could be a lot of work. Do you know of a popular arbitrary precision engine that we could potentially use here?

Reply:
The industry workhorse for arbitrary precision floats is MPFR. Berkeley's SoftFloat is not arbitrary precision, it maxes out at 128-bit IEEE floating point, with denorm and all rounding modes. BOOST has arbitrary precision floats as well.

The universal library has arbitrary floats as well, but they are not very fast. MPFR is the best solution in that regard.

Reply:
We should use arbitrary precision for constants imo, instead of using float/int literals. It would be good to pull in GMP and MPFR so we can actually represent mathematical integers and reals in the program. We did this in Lean for the AST.

Reply:
I feel like ignoring this is one of those things where people always say its not a problem until it is.

Reply:
I'm fully on board with this sentiment for thoroughness reasons. 
On the other hand, @tqchen feels that, for deep learning, we generally won't need to worry about supporting higher precision than double -- at least not immediately.
I feel, then, that the best course of action is to NOT handle arbitrary precision in this PR, but to make it a goal in the future.

Opinions?

Reply:
Oh, sorry, I see there's discussion elsewhere. Resolving this thread.